packages/@aws-cdk/core-synthesizer/lib/staging-stack.ts

```diff
@@ -0,0 +1,152 @@
+   */
+  readonly fileAssetPublishingRoleArn?: string;
+
+  /**
+   * Pass in an existing role to be used as the image publishing role.
+   *
+   * @default - a well-known name unique to this app/env.
+   */
+  readonly dockerAssetPublishingRoleArn?: string;
+}
+
+/**
+ * A default Staging Stack
+ */
+export class StagingStack extends Stack implements IStagingStack {
+  /**
+   * Default asset publishing role ARN for file (S3) assets.
+   */
+  public static readonly DEFAULT_FILE_ASSET_PUBLISHING_ROLE_ARN = 'arn:${AWS::Partition}:iam::${AWS::AccountId}:role/cdk-${Qualifier}-file-publishing-role-${AWS::AccountId}-${AWS::Region}';
+
+  /**
+   * Default asset publishing role ARN for docker (ECR) assets.
+   */
+  public static readonly DEFAULT_DOCKER_ASSET_PUBISHING_ROLE_ARN = 'cdk-${Qualifier}-asset-publishing-role-${AWS::AccountId}-${AWS::Region}';
+
+  /**
+   * The app-scoped, evironment-keyed staging bucket.
+   */
+  public readonly stagingBucket: s3.Bucket;
+
+  /**
+   * The app-scoped, environment-keyed ecr repositories associated with this app.
+   */
+  public readonly stagingRepos: Record<string, ecr.Repository>;
+
+  /**
+   * The well-known arn of the file asset publishing role.
+   */
+  public readonly fileAssetPublishingRoleArn: string;
+
+  /**
+   * The well-known arn of the docker asset publishing role.
+   */
+  public readonly dockerAssetPublishingRoleArn: string;
+
+  /**
+   * The well known name of the staging bucket
+   */
+  public readonly stagingBucketName: string;
+
+  constructor(scope: Construct, id: string, props: StagingStackProps = {}) {
+    super(scope, id, props);
+
+    this.fileAssetPublishingRoleArn = props.fileAssetPublishingRoleArn ?? StagingStack.DEFAULT_FILE_ASSET_PUBLISHING_ROLE_ARN;
+    if (!props.fileAssetPublishingRoleArn) {
+      new iam.Role(this, 'File-Asset-Publishing-Role', {
+        roleName: StagingStack.DEFAULT_FILE_ASSET_PUBLISHING_ROLE_ARN.split('/')[1],
+        assumedBy: new iam.ServicePrincipal('sts.amazonaws.com'),
+      });
+    }
+    this.dockerAssetPublishingRoleArn = props.dockerAssetPublishingRoleArn ?? StagingStack.DEFAULT_DOCKER_ASSET_PUBISHING_ROLE_ARN;
+    if (!props.dockerAssetPublishingRoleArn) {
+      new iam.Role(this, 'Docker-Asset-Publishing-Role', {
+        roleName: StagingStack.DEFAULT_DOCKER_ASSET_PUBISHING_ROLE_ARN.split('/')[1],
+        assumedBy: new iam.ServicePrincipal('sts.amazonaws.com'),
+      });
+    }
+
+    this.stagingBucketName = props.stagingBucketName ?? 'default-bucket'; // 'cdk-${AWS::AccountId}-${AWS::Region}'+`${App.of(this)?.stageName}`;
+    this.stagingBucket = new s3.Bucket(this, 'StagingBucket', {
+      bucketName: this.stagingBucketName,
+      autoDeleteObjects: true,
+      removalPolicy: RemovalPolicy.DESTROY,
+    });
+
+    this.stagingRepos = {};
+  }
+
+  public getRepoName(asset: DockerImageAssetSource): string {
+    if (!asset.uniqueId) {
+      throw new Error('uniqueId is required when using bootstrap v3');
```

Comment:
This will become a more user friendly message, right?

* Describe the problem in terms of API that the user sees (not in terms of internal code names)
* Describe the probable solution